fe/fe-core/src/main/java/org/apache/doris/datasource/iceberg/source/IcebergScanNode.java

```diff
@@ -345,23 +353,50 @@ public boolean isBatchMode() {
                 return false;
             }
         }
-        // TODO Use a better judgment method to decide whether to use batch mode.
-        return sessionVariable.getNumPartitionsInBatchMode() > 1024;
+
+        if (createTableScan().snapshot() == null) {
+            return false;
+        }
+
+        if (!sessionVariable.getEnableExternalTableBatchMode()) {
+            return false;
+        }
+
+        try (CloseableIterator<ManifestFile> matchingManifest =
+                IcebergUtils.getMatchingManifest(
+                        createTableScan().snapshot().dataManifests(icebergTable.io()),
+                        icebergTable.specs(),
+                        createTableScan().filter()).iterator()) {
+            int cnt = 0;
+            while (matchingManifest.hasNext()) {
+                cnt += matchingManifest.next().addedFilesCount();
+                if (cnt >= sessionVariable.getNumPartitionsInBatchMode()) {
+                    return true;
+                }
+            }
+        } catch (Exception e) {
+            Optional<NotSupportedException> opt = checkNotSupportedException(e);
+            if (opt.isPresent()) {
+                throw opt.get();
+            } else {
+                throw new RuntimeException(e);
+            }
+        }
+        return false;
     }
 
-    public Long getSpecifiedSnapshot() throws UserException {
+    public Long getSpecifiedSnapshot() {
         TableSnapshot tableSnapshot = getQueryTableSnapshot();
         if (tableSnapshot != null) {
             TableSnapshot.VersionType type = tableSnapshot.getType();
-            try {
-                if (type == TableSnapshot.VersionType.VERSION) {
-                    return tableSnapshot.getVersion();
-                } else {
-                    long timestamp = TimeUtils.timeStringToLong(tableSnapshot.getTime(), TimeUtils.getTimeZone());
-                    return SnapshotUtil.snapshotIdAsOfTime(icebergTable, timestamp);
+            if (type == TableSnapshot.VersionType.VERSION) {
```

Comment:
Why not wrap with `try catch`?

Reply:
Since the exception has already been handled within the `timeStringToLong` function, there will be no exception occurring here.